docs/tutorials/python/sagas/index.md

```diff
@@ -0,0 +1,551 @@
+
+    print(f"Booking car: {input.book_car_id}")
+    return f"Booked car: {input.book_car_id}"
+```
+
+The `book_hotel` and `book_flight` functions follow a similar structure:
+
+```python
+@activity.defn
+async def book_hotel(input: BookVacationInput) -> str:
+    await asyncio.sleep(3)
+    if activity.info().attempt < input.attempts:
+        activity.heartbeat(
+            f"Invoking activity, attempt number {activity.info().attempt}"
+        )
+        await asyncio.sleep(3)
+        raise RuntimeError("Hotel service is down")
+
+    if "invalid" in input.book_hotel_id:
+        raise Exception("Invalid hotel booking, rolling back!")
+
+    print(f"Booking hotel: {input.book_hotel_id}")
+    return f"Booked hotel: {input.book_hotel_id}"
+
+
+@activity.defn
+async def book_flight(input: BookVacationInput) -> str:
+    await asyncio.sleep(3)
+    if activity.info().attempt < input.attempts:
+        activity.heartbeat(
+            f"Invoking activity, attempt number {activity.info().attempt}"
+        )
+        await asyncio.sleep(3)
+        raise RuntimeError("Flight service is down")
+
+    if "invalid" in input.book_flight_id:
+        raise Exception("Invalid flight booking, rolling back!")
+
+    print(f"Booking flight: {input.book_flight_id}")
+    return f"Booked flight: {input.book_flight_id}"
+```
+
+With the main booking Activities in place, it's time to define the compensation Activities.
+These undo actions are crucial for maintaining data consistency by rolling back successful steps if a subsequent step fails.
+
+### Define Compensation Activities
+
+For every action (`book_car`, `book_hotel`, and `book_flight`), you will create a corresponding undo action.
+These Activities will log the undo action and return a success message.
+
+```python
+@activity.defn
+async def undo_book_car(input: BookVacationInput) -> str:
+    print(f"Undoing booking of car: {input.book_car_id}")
+    return f"Undoing booking of car: {input.book_car_id}"
+
+
+@activity.defn
+async def undo_book_hotel(input: BookVacationInput) -> str:
+    print(f"Undoing booking of hotel: {input.book_hotel_id}")
+    return f"Undoing booking of hotel: {input.book_hotel_id}"
+
+
+@activity.defn
+async def undo_book_flight(input: BookVacationInput) -> str:
+    print(f"Undoing booking of flight: {input.book_flight_id}")
+    return f"Undoing booking of flight: {input.book_flight_id}"
+```
+
+For this example, if the number of attempts is less than the allowed number of attempts or if the booking ID is invalid, the Activity will raise exceptions to simulate failures, then run the corresponding undo action.
+
+By setting up these compensations, you'll ensure that your system can handle failures gracefully.
+Next, you'll focus on defining shared data classes and constants to support your Activities and Workflows.
+
+## Define Shared Data Classes and Constants
+
+Shared data classes and constants are used to pass data between Activities and Workflows.
```

Comment:
Not just across Activities and Workflows though (e.g., client, workers). Maybe: "We will now create some shared data classes and constants that we will use through the application.'

Reply:
I'd like to see the title of this section and the intro paragraph make a stronger case for *why* you are doing this. This feels like something I must do for Temporal, not something that's a good practice.

Think about how you'd explain this to someone at a booth demo. Why are we doing this?